### ISSUE TITLE
Rule proposal: Avoid .map inside for loop

### ISSUE THREAD

**Issue by fregante** [sindresorhus/eslint-plugin-unicorn#846]
I've seen this a couple of times

## Error

```js 
for (const foo of bars.map(alfa => alfa.beta.gamma)) {
  foo('led again')
}
```

## Pass

```js 
for (const foo of bars) {
  foo.beta.gamma('led again')
}
```

**Comment by fisker** [comment]
I use the first one very offen, I thought It only run `.map` once? Can you explain why the first one should be avoid?

**Comment by sindresorhus** [comment]
@fisker Because you end up iterating through `bars` twice.

**Comment by fregante** [comment]
This also applies to `filter` (and `reduce`, but we already have `no-reduce` for that). I updated the examples.

I think the only case where `filter` would be useful is when the `index` is relevant, because it changes it, this would have to exclude `bars.filter(fn).entries()`

**Comment by papb** [comment]
When I first read the title, I thought the idea was to disallow it in the body of the for loop, but now I see it's not. Some name suggestions to avoid this confusion (although they don't feel perfect yet):

* `no-unnecessary-map-filter`
* `no-for-of-map-filter`

**Comment by papb** [comment]
> I think the only case where `filter` would be useful is when the `index` is relevant, because it changes it, this would have to exclude `bars.filter(fn).entries()`

I think this rule should only detect exactly `of foo.map(/* ... */)` and `of foo.filter(/* ... */)`. It should ignore any other construction because it would be too hard to statically detect whether or not it's really safe to change it...

**Comment by fregante** [comment]
> I thought the idea was to disallow it in the body of the for loop

Apparently that part is called "header", so perhaps it's `no-array-loops-in-for-header`, just in case more methods will be excluded in the future.

> It should ignore any other construction

Agreed

**Comment by papb** [comment]
Awesome, that name is much better than the ones I thought of :grinning:

**Comment by fisker** [comment]
> @fisker Because you end up iterating through `bars` twice.

Oh, right.  I see.

In that case 

```js
cont foo = bar.map(fn)
for (const gamma of foo) {
  gamma('ray')
}
```

the same thing, it's like do not use `.map/filter` if you are going to run `for-of` on the array.

**Comment by fregante** [comment]
> the same thing

That's right! Is that as easily detectable? Can the rule ensure that `foo` isn't also used later?

If so, the rule could be more generically `no-duplicate-loops`

**Comment by papb** [comment]
@fregante Your last example can't be safely converted to a single loop... (in general)

**Comment by fregante** [comment]
> Your last example can't be safely converted to a single loop

If my last example isn't, then none of them is.

If auto fix isn't safe, then just the auto fix can't be implemented.

**Comment by fisker** [comment]
> Is that as easily detectable? 

Maybe not.

**Comment by fisker** [comment]
Should we consider this edge case

```
for (const gamma of 
  bars.map(someFunctionDoSomethingOrChangesThePreviousElement)
) {
  gamma('ray')
}
```

**Comment by futpib** [comment]
How about forcing the looped expression to be one of:
* a variable (`x`)
* a property access (`x.a.b`)
* a single function call with no arguments (`x()`, `x.a.b()`)
* a well-known iterator-returning method or built-in function call:
* * `x.keys()`/`x.values()`/`x.entries()`
* * `Object.keys(x)`/`Object.values(x)`/`Object.entries(x)`

Otherwise force the user to define a variable before the loop (may be possible to auto-fix, the iteratee variable name could be the pluralized name of the element variable).

**Comment by papb** [comment]
> 
> 
> Should we consider this edge case
> 
> ```
> for (const gamma of 
>   bars.map(someFunctionDoSomethingOrChangesThePreviousElement)
> ) {
>   gamma('ray')
> }
> ```

Hmm, I think functions with side-effects inside `map` and `filter` are incorrect anyway, so if that gives a lint error that is fine by me. Maybe the error message can be:

> Do not use `map` in the header of a `for` loop. Perform the operation directly inside the loop instead. Note that your mapping must not have side-effects.

> Do not use `filter` in the header of a `for` loop. Perform an `if` directly inside the loop instead. Note that your filter predicate must not have side-effects.

**Comment by papb** [comment]
@futpib Hmm I don't like that limitation, I can see myself using `x().y()` for example, or a function with a short parameter, `x.foo(bar)`.

**Comment by sindresorhus** [comment]
> @fregante Your last example can't be safely converted to a single loop... (in general)

"in general", really? I would argue in general it would be safe, but in some cases, the iterator might be single-pass or there are mutations between the loops. In any case, I think we could detect some cases of mutations between the loop statements, and we would of course not auto-fix this case.

**Comment by sindresorhus** [comment]
> no-duplicate-loops

I would call it this. We can start with the simple cases and then open an issue for follow-up (advanced) improvements.

**Comment by sindresorhus** [comment]
> How about forcing the looped expression to be one of:

That sounds sensible, but wouldn't that be a completely different rule? This rule is about preventing duplicate iterations.

**Comment by sindresorhus** [comment]
> Should we consider this edge case

It's clear that auto-fixing is unsafe, but I think we shouldn't care about edge-cases like this for reporting.

**Comment by sindresorhus** [comment]
@futpib Can you open a separate proposal for https://github.com/sindresorhus/eslint-plugin-unicorn/issues/846#issuecomment-703317406 when you have time?

**Comment by sindresorhus** [comment]
This is now accepted. Pull request welcome.

**Comment by manovotny** [comment]
I was going to create another new rule proposal for the following but I wonder if it fits within this rule or maybe in #1044.

I have been in a couple of code bases recently where I've seen something like the following:

```
videos
    .filter((video) => !video.expired)
    .filter((video) => video.season > 0)
    .filter((video) => video.episode > 0)
```

This is iterating over the same array three times when it could be done in a single iteration:

```
videos.filter((video) => !video.expired && video.season > 0 && video.episode > 0)
```

@sindresorhus @fisker thoughts on this also being included in this same rule or does it deserve it's own rule (something like `no-consecutive-filter`, `no-chained-filter`, or change `filter` to `iteration` if you want to go after other chains like `.map`, etc.)?

**Comment by papb** [comment]
> > @fregante Your last example can't be safely converted to a single loop... (in general)
> 
> "in general", really? I would argue in general it would be safe, but in some cases, the iterator might be single-pass or there are mutations between the loops. In any case, I think we could detect some cases of mutations between the loop statements, and we would of course not auto-fix this case.

Yes, I could have said it better. What I wanted to say is that merging the two loops is not always correct.

```
for (const gamma of bars) {
  gamma('ray')
}
for (const gamma of bars) {
  doSomethingWith(gamma)
}
```

For consecutive loops as above, I'd suggest the rule to give a warning (not error!) and a message like:

> Check if you can merge the two loops into one, and if not, disable this warning with `// eslint-disable-next-line no-duplicate-loops`

**Comment by omril1** [comment]
I've seen another example that got highlighted to me from a lint fix of `unicorn/prefer-spread`:

```js
for (const child of Array.from(currentElement.children)) {
  queue.push([child, currentDepth + 1]);
}
```

The fixer changed it to
```js
for (const child of [...currentElement.children]) {
    queue.push([child, currentDepth + 1]);
}
```
Which made it more clear that iterates twice, I think both should be reported as for `no-duplicate-loops`
